File: scripts/discogs/export.py
#!/usr/bin/env python3

import sys
import os
import json
from urllib import request, parse

def fetch_releases(token, page = 1):
    discogs_base_url = "https://api.discogs.com"

    url = discogs_base_url + "/users/ngalaiko/collection/folders/0/releases?sort=artist&page={}".format(page)
    req = request.Request(url, headers={
        "Content-Type": "application/json",
        "Authorization": "Discogs token={}".format(token),
    })
    resp = request.urlopen(req).read()
    releases_resp = json.loads(resp)
    releases = releases_resp['releases']

    pagination = releases_resp['pagination']
    if not 'next' in pagination['urls']:
        return releases

    return releases + fetch_releases(token, page=page+1)

def save_releases(releases):
    script_location = os.path.realpath(__file__)
    script_dir = os.path.dirname(script_location)
    dst = os.path.join(script_dir, "../../data/records.json")
    dst = os.path.normpath(dst)

    content = json.dumps({'releases': releases}, sort_keys=True, indent=2)
    with open(dst, 'w') as f:
        f.write(content)

if __name__ == "__main__":
    if len(sys.argv) != 2: 
        print("usage: {} <discogs_token>".format(sys.argv[0]))
        sys.exit(1)

    token = sys.argv[1]
    releases = fetch_releases(token)
    save_releases(releases)

File: scripts/discogs/export.ts
import yargs from 'yargs';
import fetch from 'node-fetch';
import { writeFile } from 'fs';

const argv = yargs(process.argv.slice(2))
	.usage('Usage: $0 <command> [options]')
	.option('api-token', {
		alias: 't',
		describe: 'The token to use for authentication',
		type: 'string',
		demandOption: true
	})
	.option('output', {
		alias: 'o',
		describe: 'The output file',
		type: 'string',
		demandOption: true
	});

const download = async (token: string, page = 1) =>
	fetch(
		`https://api.discogs.com/users/ngalaiko/collection/folders/0/releases?sort=artist&page=${page}`,
		{
			headers: {
				Authorization: `Discogs token=${token}`,
				Accept: 'application/json'
			}
		}
	)
		.then((res) => res.json())
		.then(async (json: any) => {
			const releases = json.releases;
			const { next } = json.pagination.urls;
			return next ? [...releases, ...(await download(token, page + 1))] : releases;
		});

const writeJSON =
	(path: string) =>
	(data: any): Promise<void> =>
		new Promise((resolve, reject) => {
			console.log('writing to', path);
			writeFile(path, JSON.stringify(data), (error) => {
				if (error) {
					reject(error);
				} else {
					resolve();
				}
			});
		});

await download(argv.argv.apiToken).then(writeJSON(argv.argv.output));
